packages/cli/src/lib/oauth/oauthClientServer.ts

```diff
@@ -0,0 +1,102 @@
+import axios, {AxiosRequestConfig} from 'axios';
+import {createServer, IncomingMessage, ServerResponse} from 'http';
+import {URLSearchParams} from 'url';
+import {AuthorizationError, InvalidStateError} from './authorizationError';
+import {AuthorizationServiceConfiguration, ClientConfig} from './oauthConfig';
+
+export class OAuthClientServer {
+  public constructor(
+    private clientConfig: ClientConfig,
+    private authServiceConfig: AuthorizationServiceConfiguration
+  ) {}
+
+  public startServer(port: number, hostname: string, expectedState: string) {
+    return new Promise<{accessToken: string}>((resolve, reject) => {
+      const listener = async (req: IncomingMessage, res: ServerResponse) => {
+        try {
+          const {code, state} = this.parseUrl(req);
+          const {tokenEndpoint} = this.authServiceConfig;
+
+          if (!state || !code) {
+            return;
+          }
+
+          if (state !== expectedState) {
+            throw new InvalidStateError(state, expectedState);
+          }
+
+          const data = this.getTokenQueryString(code);
+          const authRequest = await axios({
+            method: 'post',
+            url: tokenEndpoint,
+            ...this.requestConfig,
+            data,
+          });
+          const accessToken = authRequest.data.access_token;
+
+          res.end('Close your browser to continue');
+          resolve({accessToken});
+          server.close();
```

Comment:
Those `close()` call should be handled in some kind of `finally`, because it does not close when we throw on line 25, or for any other random error that could happen, right ?

Reply:
Yes, it "should" close the server even if an error is fired.
Any error will be intercepted in the `catch` block which will close the server and gracefully return the error message to be handled by Oclif.

![image](https://user-images.githubusercontent.com/12199712/135912310-6d6d7a14-c547-4b69-9a54-e86f1abcaead.png)

Reply:
It's true that line 42 we do call server.close but @olamothe ain't wrong to say we should do the `server.close()` in a `finally`:
 - It avoid repetition (one line, I admit, that's nothing)
 - More importantly, it's semantically more accurate: "we do this, if things goes haywire we do catch, and no matter what we finish by closing the server"